File: LeetCode/cpp/1743_Restore_the_Array_From_Adjacent_Pairs.cpp
class Solution {
public:
	unordered_map<int, vector<int>> makeMap(vector<vector<int>>& adjacentPairs) {
		unordered_map<int, vector<int>> um;

		for (vector<int> pairs : adjacentPairs) {
			int a = pairs[0];
			int b = pairs[1];

			um[a].push_back(b);
			um[b].push_back(a);
		}

		return um;
	}

	unordered_map<int, int> makeCntMap(vector<vector<int>>& adjacentPairs) {
		unordered_map<int, int> um;

		for (vector<int> pairs : adjacentPairs) {
			int a = pairs[0];
			int b = pairs[1];

			um[a]++;
			um[b]++;
		}

		return um;
	}

	set<int> initializeVisit(vector<vector<int>>& adjacentPairs) {
		set<int> visit;

		for (vector<int> pairs : adjacentPairs) {
			int a = pairs[0];
			int b = pairs[1];

			visit.insert(a);
			visit.insert(b);
		}

		return visit;
	}

	pair<int, int> findOneCountNumber(unordered_map<int, int>& cntMap) {
		pair<int, int> pair = { INT_MAX, INT_MAX };

		for (auto m : cntMap) {
			if (m.second == 1) {
				if (pair.first == INT_MAX) {
					pair = { m.first, INT_MAX };
				}
				else {
					pair = { pair.first, m.first };
				}
			}
		}

		return pair;
	}

	vector<int> bfs(int start, unordered_map<int, vector<int>>& um, unordered_map<int, int>& cntMap, set<int>& visit) {
		vector<int> answer;

		queue<int> q;
		q.push(start);
		visit.erase(start);

		while (!q.empty()) {
			int x = q.front();
			q.pop();

			answer.push_back(x);

			if (um.count(x) == 1) {
				vector<int> v = um[x];

				for (int i = 0; i < v.size(); i++) {
					if (cntMap[v[i]] == 2 && visit.find(v[i]) != visit.end()) {
						q.push(v[i]);
						visit.erase(v[i]);
					}
				}
			}
		}

		return answer;
	}

	vector<int> restoreArray(vector<vector<int>>& adjacentPairs) {
		vector<int> answer;

		unordered_map<int, vector<int>> um = makeMap(adjacentPairs);
		unordered_map<int, int> cntMap = makeCntMap(adjacentPairs);

		set<int> visit = initializeVisit(adjacentPairs);

		pair<int, int> oneCnt = findOneCountNumber(cntMap);

		answer = bfs(oneCnt.first, um, cntMap, visit);

		answer.push_back(oneCnt.second);

		return answer;
	}
};

File: LeetCode/java/1743_Restore_the_Array_From_Adjacent_Pairs.java
import java.util.*;

class Solution {
    private Map<Integer, List<Integer>> makeMap(int[][] adjacentPairs) {
        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int[] pairs : adjacentPairs) {
            int a = pairs[0];
            int b = pairs[1];

            if (!map.containsKey(a)) {
                map.put(a, new ArrayList<>());
            }

            map.get(a).add(b);

            if (!map.containsKey(b)) {
                map.put(b, new ArrayList<>());
            }

            map.get(b).add(a);
        }

        return map;
    }

    private Map<Integer, Integer> makeCntMap(int[][] adjacentPairs) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int[] pairs : adjacentPairs) {
            int a = pairs[0];
            int b = pairs[1];

            map.put(a, map.getOrDefault(a, 0) + 1);
            map.put(b, map.getOrDefault(b, 0) + 1);
        }

        return map;
    }

    private Set<Integer> initializeVisit(int[][] adjacentPairs) {
        Set<Integer> visit = new HashSet<>();

        for (int[] pairs : adjacentPairs) {
            int a = pairs[0];
            int b = pairs[1];

            visit.add(a);
            visit.add(b);
        }

        return visit;
    }

    private int[] findOneCountNumber(Map<Integer, Integer> cntMap) {
        int[] pair = new int[] {Integer.MAX_VALUE, Integer.MAX_VALUE};

        for (Map.Entry<Integer, Integer> map : cntMap.entrySet()) {
            if (map.getValue() == 1) {
                if (pair[0] == Integer.MAX_VALUE) {
                    pair = new int[] {map.getKey(), Integer.MAX_VALUE};
                } else {
                    pair = new int[] {pair[0], map.getKey()};
                }
            }
        }

        return pair;
    }

    private int[] bfs(int start, Map<Integer, List<Integer>> map, Map<Integer, Integer> cntMap, Set<Integer> visit) {
        int[] answer = new int[map.size()];
        int idx = 0;

        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        visit.remove(start);

        while (!q.isEmpty()) {
            int x = q.poll();

            answer[idx++] = x;

            if (map.containsKey(x)) {
                List<Integer> list = map.get(x);

                for (int i = 0; i < list.size(); i++) {
                    if (cntMap.get(list.get(i)) == 2 && visit.contains(list.get(i))) {
                        q.add(list.get(i));
                        visit.remove(list.get(i));
                    }
                }
            }
        }

        return answer;
    }

    public int[] restoreArray(int[][] adjacentPairs) {
        int[] answer;

        Map<Integer, List<Integer>> map = makeMap(adjacentPairs);
        Map<Integer, Integer> cntMap = makeCntMap(adjacentPairs);

        Set<Integer> visit = initializeVisit(adjacentPairs);

        int[] oneCnt = findOneCountNumber(cntMap);

        answer = bfs(oneCnt[0], map, cntMap, visit);
        answer[answer.length - 1] = oneCnt[1];

        return answer;
    }
}